docs: adds enterprise documentation for plugin wif (#25706)
hashicorp/vault: website/content/docs/secrets/aws.mdx

    on using STS Federation Tokens, since the temporary security credentials
    associated with the role are not authorized to use GetFederationToken.)

>     In some cases, providing sensitive IAM security credentials to Vault as configuration 
>     is not possible. For example, your organization may require that all security 
>     credentials are short-lived or tied to a machine identity. We recommend using Vault 
>     [plugin workload identity federation](#plugin-workload-identity-federation-wif) (WIF) 
>     in this case.
> 
>     See the IAM [permissions policy](#iam-permissions-policy-for-vault) for Vault section 
>     for the permissions the credentials need to have for Vault to perform its actions.

Reviewer: ```suggestion
    In some cases, you cannot set sensitive IAM security credentials in your
    Vault configuration. For example, your organization may require that all
    security credentials are short-lived or explicitly tied to a machine identity.
    
    To provide IAM security credentials outside your Vault configuration, we
    recommend using Vault
    [plugin workload identity federation](#plugin-workload-identity-federation-wif)
    (WIF).

    Refer to the IAM [permissions policy for Vault](#iam-permissions-policy-for-vault)
    section for details on the permission requirements for using WIF.
```

Author: > To provide IAM security credentials outside your Vault configuration

There aren't IAM security credentials being provided outside of the Vault configuration. They aren't being provided at all in the case of plugin WIF.

> Refer to the IAM [permissions policy for Vault](#iam-permissions-policy-for-vault) section for details on the permission requirements for using WIF.

These are the general permissions needed regardless of using plugin WIF or not.

Happy to commit a different suggestion or commit this one followed by edits!

Reviewer: Tweaked the suggestion based on your clarifications. Given that the permissions are general, and not WIF-specific, I don't think it adds much to call them out here. If they aren't being highlighted effectively elsewhere in the doc, we can add them as a suggested reading at the end

Author: Makes sense, thanks!

ensure that Vault uses one of the permissions boundaries specified (not all of
them).

> ### Plugin Workload Identity Federation (WIF)

Reviewer 2: Is this intentionally a subsection of "IAM permissions policy for Vault"? It seems like it shouldn't be.

Author: Good catch! Not intentional. Moved in https://github.com/hashicorp/vault/pull/25706/commits/a4f5a24251cfbe16cd566ee906d549ba826208aa.

ensure that Vault uses one of the permissions boundaries specified (not all of
them).

### Plugin Workload Identity Federation (WIF)

<EnterpriseAlert product="vault" />

> The AWS secrets engine supports the Plugin WIF workflow, and has a source of identity called
> a plugin identity token. The plugin identity token is a JWT that is internally signed by Vault's 
> [plugin identity token issuer](/vault/api-docs/secret/identity/tokens#read-plugin-workload-identity-issuer-s-openid-configuration). 
> Given a trust relationship is configured between Vault and AWS via [Web Identity Federation](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_oidc.html),
> the secrets engine can exchange its identity token for short-lived STS credentials needed to
> perform its actions. This enables the secrets engine to operate without needing sensitive
> IAM security credentials as a part of its configuration.

Reviewer: ```suggestion
The AWS secrets engine supports the Plugin WIF workflow, and has a source of identity called
a plugin identity token. The plugin identity token is a JWT that is internally signed by Vault's 
[plugin identity token issuer](/vault/api-docs/secret/identity/tokens#read-plugin-workload-identity-issuer-s-openid-configuration). 

If there is a trust relationship configured between Vault and AWS through
[Web Identity Federation](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_oidc.html),
the secrets engine can exchange its identity token for short-lived STS credentials needed to
perform its actions.

Exchanging identity tokens for STS credentials lets the AWS secrets engine
operate without configuring explicit access to sensitive IAM security
credentials.
```

Author: Thanks 👍

> To configure the secrets engine to use plugin WIF, you must first:

Reviewer: ```suggestion
To configure the secrets engine to use plugin WIF:
```

Author: Thanks!

1. Ensure that Vault [openid-configuration](/vault/api-docs/secret/identity/tokens#read-plugin-identity-token-issuer-s-openid-configuration)
   and [public JWKS](/vault/api-docs/secret/identity/tokens#read-plugin-identity-token-issuer-s-public-jwks) 

>    APIs are network-reachable by AWS
>    
>    An API proxy or gateway can be used to limit Vault API exposure depending on your environment.  

Reviewer: ```suggestion
   APIs are network-reachable by AWS. We recommend using an API proxy or gateway
   if you need to limit Vault API exposure.  
```

Author: Thanks!

> 1. Create an [IAM OIDC identity provider](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html) in AWS
>    
>    The provider URL must point at Vault's [plugin identity token issuer](/vault/api-docs/secret/identity/tokens#read-plugin-workload-identity-issuer-s-openid-configuration). 
>    Remove the `/.well-known/openid-configuration` suffix of the URL and provide it like 
>    `https://host:port/v1/identity/oidc/plugins`.
>    
>    The audience should uniquely identify the recipient of the plugin identity token. In AWS, 
>    the recipient is the identity provider. We recommend setting this to the `host:port/v1/identity/oidc/plugins`
>    portion of the provider URL since it's enforced as unique per configured identity provider.

Reviewer: ```suggestion
1. Create an
   [IAM OIDC identity provider](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html)
   in AWS.
   1. The provider URL **must** point at your [Vault plugin identity token issuer](/vault/api-docs/secret/identity/tokens#read-plugin-workload-identity-issuer-s-openid-configuration) with the
   `/.well-known/openid-configuration` suffix removed. For example: 
   `https://host:port/v1/identity/oidc/plugins`.
   1. The audience should uniquely identify the recipient of the plugin identity
   token. In AWS, the recipient is the identity provider. We recommend using
   the `host:port/v1/identity/oidc/plugins` portion of the provider URL as your
   recipient since it will be unique for each configured identity provider.
```

Author: Thanks!

> 1. Create a [web identity role](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-idp_oidc.html#idp_oidc_Create) in AWS
>    
>    Select the identity provider and audience created in step 2 when configuring the role.
> 
> Next, configure the AWS secrets engine with the audience value and role ARN created from steps 2 and 3:

Reviewer: ```suggestion
1. Create a [web identity role](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-idp_oidc.html#idp_oidc_Create)
   in AWS with the same audience used for your IAM OIDC identity provider.
1. Configure the AWS secrets engine with the IAM OIDC audience value and web
   identity role ARN.
```

Author: Thanks!

```shell-session
$ vault write aws/config/root \
    identity_token_audience="vault.example/v1/identity/oidc/plugins" \
    role_arn="arn:aws:iam::123456789123:role/example-web-identity-role"
```

> The secrets engine is now configured to use plugin WIF for its configuration credentials. 
> These [credentials](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRoleWithWebIdentity.html) 
> are short-lived (default 1hr) and will be automatically refreshed through expiration without 
> need for manual intervention.
> 
> Please see the [API documentation](/vault/api-docs/secret/aws#configure-root-credentials) 
> for more details on the fields associated with plugin WIF.

Reviewer: ```suggestion
Your secrets engine can now use plugin WIF for its configuration credentials. 
By default WIF [credentials](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRoleWithWebIdentity.html) 
have a time-to-live of 1 hour and automatically refresh when they expire.

Please see the [API documentation](/vault/api-docs/secret/aws#configure-root-credentials) 
for more details on the fields associated with plugin WIF.
```
